src/machine/operation.c

```diff
@@ -8,8 +8,33 @@
 #include "operation.h"
 #include "process-util.h"
 
+static int operation_done_internal(const siginfo_t *si, Operation *o, void *userdata) {
+        int r;
+
+        assert(si);
+        assert(o);
+
+        if (si->si_code != CLD_EXITED)
+                return log_debug_errno(SYNTHETIC_ERRNO(ESHUTDOWN), "Child died abnormally");
+
+        if (si->si_status == EXIT_SUCCESS)
+                r = 0;
+        else if (read(o->errno_fd, &r, sizeof(r)) != sizeof(r)) /* Try to acquire error code for failed operation */
+                return log_debug_errno(errno, "Failed to read failed operation's errno: %m");
+
+        if (o->done)
+                /* A completion routine is set for this operation, call it. */
+                return o->done(o, r, userdata);
+
+        /* The default operation when done is to simply return an error on failure or an empty success
+         * message on success. */
+        if (r < 0)
+                log_debug_errno(r, "Failed operation's errno: %m");
```

Comment:
The log message sounds spurious for me. How about simply "Operation failed: %m" or so?